variants/backend-base/lib/tasks/dev.rake

```diff
@@ -0,0 +1,11 @@
+##
+# The dev namespace is for tasks which are intended for local development only.
+#
+# If your task could be run in local development or a deployed environment you
```

Comment:
I think it would be better to say something like "is suitable for" because I expect you "could" run most custom rake tasks in any environment, but it doesn't mean you should.

tbh I actually think it would be good to stick a guard at the very top of the file that returns if you're not in `development` mode:

```
return unless ENV["RAILS_ENV"] == "development"

require "aws_ec2_environment"
require "aws-sdk-ec2"

namespace :aws do # rubocop:disable Metrics/BlockLength
  namespace :ec2 do # rubocop:disable Metrics/BlockLength
    desc "Lists all the running EC2 instances in the configured account & region"
    task list_running_instances: :environment do # rubocop:disable Metrics/BlockLength
      # ...
    end
  end
end
```

Testing that locally it seems to work:

```
rails-template-example-aws on  ssm [!?] via  v16.15.0 via 💎 v3.1.2
❯ bundle exec rake -T | grep aws
rake aws:ec2:list_running_instances                            # Lists all the running EC2 instances in the configured account & region
rake aws:ssm:forward_port[instance_id,remote_port,local_port]  # Starts a port forwarding session to the given instance, even if the instance is in a private subnet

rails-template-example-aws on  ssm [!?] via  v16.15.0 via 💎 v3.1.2
❯ RAILS_ENV=production bundle exec rake -T | grep aws
```

I see two advantages to that:

1. enforces that these tasks really are only for local dev, and helps guards against people doing stuff like `dev:seed`
   - and if someone had a really good reason for wanting to run such a task on a deployed env, I think it's reasonable to expect them to know how to bypass that guard safely anyway
2. doesn't result in require calls that could be for gems that are only installed in `development`, which won't be installed on deployed environments so will break `rake` otherwise

Reply:
I'm not in favour of a guard at the top of the file because:

* I rely on `rake -T` to tell me what's available on an app so having it give different answers in different envs feels like a source of unnecessary confusion
* Files in `lib/` are not autoloaded so any required gems which only exist in development will only break the file they are required in. This seems like a totally fine outcome tbh.

In general, I feel like this is a situation where we want to "encourage" behaviour rather than "enforce" it. I have had instances where I definitely did want to run developer seeds in a deployed env (e.g. setting up a new pre-prod env in Heroku) and more generally when I ended up needing to run some developer focused rake task in a pre-prod env. We want to encourage devs to put those tasks in `app.rake` but I don't think I want to punish them if they didn't predict the future properly

Reply:
> Files in lib/ are not autoloaded so any required gems which only exist in development will only break the file they are required in. This seems like a totally fine outcome tbh.

That's not correct:

```
deploy@ip-10-0-57-29:~/rails-app-standard-example/current$ bundle exec rake -T
rake aborted!
LoadError: cannot load such file -- aws-sdk-ec2
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bootsnap-1.13.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:29:in `require'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require'
/home/deploy/rails-app-standard-example/releases/20221015204246/lib/tasks/aws.rake:1:in `<main>'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/engine.rb:661:in `load'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/engine.rb:661:in `block in run_tasks_blocks'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/engine.rb:661:in `each'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/engine.rb:661:in `run_tasks_blocks'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/application.rb:506:in `run_tasks_blocks'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/engine.rb:464:in `load_tasks'
/home/deploy/rails-app-standard-example/releases/20221015204246/Rakefile:6:in `<top (required)>'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli/exec.rb:58:in `load'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli/exec.rb:58:in `kernel_load'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli/exec.rb:23:in `run'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli.rb:483:in `exec'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli.rb:31:in `dispatch'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli.rb:25:in `start'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/exe/bundle:48:in `block in <top (required)>'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/friendly_errors.rb:120:in `with_friendly_errors'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/exe/bundle:36:in `<top (required)>'
/usr/lib/rbenv/versions/3.1.2/bin/bundle:25:in `load'
/usr/lib/rbenv/versions/3.1.2/bin/bundle:25:in `<main>'
(See full trace by running task with --trace)
deploy@ip-10-0-57-29:~/rails-app-standard-example/current$ bundle exec rake app:say_hello
rake aborted!
LoadError: cannot load such file -- aws-sdk-ec2
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bootsnap-1.13.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:29:in `require'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require'
/home/deploy/rails-app-standard-example/releases/20221015204246/lib/tasks/aws.rake:1:in `<main>'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/engine.rb:661:in `load'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/engine.rb:661:in `block in run_tasks_blocks'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/engine.rb:661:in `each'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/engine.rb:661:in `run_tasks_blocks'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/application.rb:506:in `run_tasks_blocks'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/railties-7.0.3.1/lib/rails/engine.rb:464:in `load_tasks'
/home/deploy/rails-app-standard-example/releases/20221015204246/Rakefile:6:in `<top (required)>'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli/exec.rb:58:in `load'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli/exec.rb:58:in `kernel_load'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli/exec.rb:23:in `run'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli.rb:483:in `exec'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli.rb:31:in `dispatch'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/cli.rb:25:in `start'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/exe/bundle:48:in `block in <top (required)>'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/lib/bundler/friendly_errors.rb:120:in `with_friendly_errors'
/home/deploy/rails-app-standard-example/shared/bundle/ruby/3.1.0/gems/bundler-2.3.17/exe/bundle:36:in `<top (required)>'
/usr/lib/rbenv/versions/3.1.2/bin/bundle:25:in `load'
/usr/lib/rbenv/versions/3.1.2/bin/bundle:25:in `<main>'
(See full trace by running task with --trace)
```

This error is because I've got `lib/tasks/aws.rake` which uses the `aws-sdk-ec2` gem, which is only included in the `development` group:

```
deploy@ip-10-0-57-29:~/rails-app-standard-example/current$ cat lib/tasks/app.rake
namespace :app do
  desc "My app-related task"
  task say_hello: :environment do
    puts "hello world!"
  end
end
deploy@ip-10-0-57-29:~/rails-app-standard-example/current$ head lib/tasks/aws.rake
require "aws-sdk-ec2"

namespace :aws do # rubocop:disable Metrics/BlockLength
  namespace :ec2 do # rubocop:disable Metrics/BlockLength
    desc "Lists all the running EC2 instances in the configured account & region"
    task list_running_instances: :environment do # rubocop:disable Metrics/BlockLength
      ec2 = Aws::EC2::Client.new

      instances = ec2.describe_instances(
        filters: [{ name: "instance-state-name", values: ["running"] }]
```

> I rely on rake -T to tell me what's available on an app so having it give different answers in different envs feels like a source of unnecessary confusion

I don't disagree, but at the same time it's actually still accurate: `rake -T` would be showing you what's available _for the environment you're running in_ (it would be nice if that could be surfaced more clearly though)

Reply:
`rescue LoadError` or `require 'aws-sdk-ec2' if Rails.env.development?` or this feels like a reasonable reason to have the requires inside the tasks (which i'm not opposed to at all)

there are ways to have your cake (comprehensive rake -T) and eat it to (not have all gems on production)

Reply:
I'm not opposed to those, but they feel like they're still worse evils to me because it means we're listing tasks that we know will explode if they're used, and anything we do to try and improve that (i.e. having a helper that prints an explainer) adds more complexity that devs need to know about here which differs from their usual flow; and that's not something we're not always going to catch in our current setup e.g. in my example, `aws-sdk-ec2` is needed for deployments so it's present in both `development` and `test` so will be installed in CI - so you don't see the error until you try and run a rake task on the server (it might end up exploding when deploying due to doing `assets precompile`, but I don't think we should rely on that).

I just think it's easier to have my suggested guard and explain it as "`rake -T` only shows tasks that are suitable for the environment you're running in" 🤷

Reply:
I think most Rails devs would never think to check that `rake -T` was per-environment (even if we put it in the README). FWIW I think putting the `require` at the top of the task feels fine to me.

In my experience, it's fairly unusual to have rake tasks which have gems which work in dev only - I'm not saying there is anything wrong with it or that it's a bad choice for your `aws.rake` stuff, just that it's not the common case so I don't think this will set a bad example for devs in terms of how they use `require`.

Reply:
btw this is how the [annotate gem](https://github.com/ctran/annotate_models) handles this for the rake task it creates which runs annotations after migrations

```ruby

# NOTE: only doing this in development as some production environments (Heroku)
# NOTE: are sensitive to local FS writes, and besides -- it's just not proper
# NOTE: to have a dev-mode tool do its thing in production.
if Rails.env.development?
  require 'annotate'
  task :set_annotation_options do
    # You can override any of these by setting an environment variable of the
    # same name.
    Annotate.set_defaults(
      'active_admin'                => 'false',
      'additional_file_patterns'    => [],
      'routes'                      => 'false',
      'models'                      => 'true',
      'position_in_routes'          => 'before',
      'position_in_class'           => 'before',
# snip
```
This might be a clean solution for your aws stuff @G-Rath

Reply:
TIL about rake loading code from all tasks when you run a task. Obviously my mental model of `lib/` not being autoloaded wasn't nuanced enough. I'm pretty sure it's not autoloaded by Rails but obviously not rake. Thanks @G-Rath

Reply:
@eoinkelly I'm not sure what you mean - that looks like the same as what I originally suggested, except they're wrapping everything in the guard instead of returning early? (which has the same result, just different levels of indenting).

Looking in their codebase, for their built-in tasks they do something similar again except instead of just returning the define the task with no body.

Reply:
Discussed in meeting. Resolution was to include the guard in a comment explaining what you might hit if you require dev only gems in prod. Also include other solutions in the comment e.g. require within task

Reply:
I have updated the comment explaining the situation - this is ready for review again.